test/FlashLoan.t.sol

```diff
@@ -0,0 +1,279 @@
+            bidAmount: 1e18,
+            value: 10e18
+        });
+
+        (sig.v, sig.r, sig.s) = vm.sign(solverOnePK, atlasVerification.getSolverPayload(solverOps[0]));
+        solverOps[0].signature = abi.encodePacked(sig.r, sig.s, sig.v);
+
+        dAppOp = txBuilder.buildDAppOperation(governanceEOA, userOp, solverOps);
+        (sig.v, sig.r, sig.s) = vm.sign(governancePK, atlasVerification.getDAppOperationPayload(dAppOp));
+        dAppOp.signature = abi.encodePacked(sig.r, sig.s, sig.v);
+
+        // Call again with partial payback, should still revert
+        vm.startPrank(userEOA);
+        vm.expectEmit(true, true, true, true);
+        emit FastLaneErrorsEvents.SolverTxResult(address(solver), solverOneEOA, true, false, 8388610);
+        vm.expectRevert();
+        atlas.metacall({ userOp: userOp, solverOps: solverOps, dAppOp: dAppOp });
+        vm.stopPrank();
+
+        // final try, should be successful with full payback
+        solverOps[0] = txBuilder.buildSolverOperation({
+            userOp: userOp,
+            solverOpData: abi.encodeWithSelector(SimpleSolver.payback.selector, 1e18),
+            solverEOA: solverOneEOA,
+            solverContract: address(solver),
+            bidAmount: 1e18,
+            value: 10e18
+        });
+
+        (sig.v, sig.r, sig.s) = vm.sign(solverOnePK, atlasVerification.getSolverPayload(solverOps[0]));
+        solverOps[0].signature = abi.encodePacked(sig.r, sig.s, sig.v);
+
+        dAppOp = txBuilder.buildDAppOperation(governanceEOA, userOp, solverOps);
+        (sig.v, sig.r, sig.s) = vm.sign(governancePK, atlasVerification.getDAppOperationPayload(dAppOp));
+        dAppOp.signature = abi.encodePacked(sig.r, sig.s, sig.v);
+
+        uint256 solverStartingWETH = WETH.balanceOf(address(solver));
+        uint256 atlasStartingETH = address(atlas).balance;
+        uint256 userStartingETH = address(userEOA).balance;
+
+        assertEq(solverStartingWETH, 1e18, "solver incorrect starting WETH");
+        assertEq(atlasStartingETH, 102e18, "atlas incorrect starting ETH"); // 2e initial + 1e solver + 100e user deposit
+
+        // Last call - should succeed
+        vm.startPrank(userEOA);
+        atlas.metacall({ userOp: userOp, solverOps: solverOps, dAppOp: dAppOp });
+        vm.stopPrank();
+
+        uint256 solverEndingWETH = WETH.balanceOf(address(solver));
+        uint256 atlasEndingETH = address(atlas).balance;
+        uint256 userEndingETH = address(userEOA).balance;
+
+        console.log("solverWETH", solverStartingWETH, solverEndingWETH);
+        console.log("atlasETH", atlasStartingETH, atlasEndingETH);
+        console.log("userETH", userStartingETH, userEndingETH);
+
+        assertEq(solverEndingWETH, 0, "solver WETH not used");
+        // assertEq(atlasEndingETH - atlasStartingETH, 999612266500000000, "atlas incorrect ending ETH"); // atlas should receive bid
+
+    }
+}
+
+contract DummyController is DAppControl {
```

Comment:
Can use the the `DummerDAppControlBuilder`, especially if the controller doesn't do anything special.